### PR TITLE
[firebase_ml_vision] Firebase AutoML Vision Edge Support

### PR DESCRIPTION
**Pull request by rishab2113**
## Description

Added support to perform image labeling using Local and Remote Models generated using AutoML Vision. Added because the package didn't have it yet, new product release at Google I/O.

## Related Issues

None Known

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

**Comment by Baileypollard** [comment]
@bparrishMines Can we get a code review on this? This is a important feature that would be very useful to us.

File: packages/firebase_ml_vision/android/src/main/java/io/flutter/plugins/firebasemlvision/RemoteVisionEdgeDetector.java [hunk 1]
```diff
@@ -0,0 +1,163 @@
+package io.flutter.plugins.firebasemlvision;
```
Comment (bparrishMines) [inline packages/firebase_ml_vision/android/src/main/java/io/flutter/plugins/firebasemlvision/RemoteVisionEdgeDetector.java]: Same as the file above.

File: packages/firebase_ml_vision/example/android/app/build.gradle [hunk 2]
```diff
@@ -31,9 +31,13 @@ android {
...
     }
 
+    aaptOptions {
+        noCompress "tflite"
+    }
+
     defaultConfig {
         applicationId "io.flutter.plugins.firebasemlvisionexample"
-        minSdkVersion 16
+        minSdkVersion 21
```
Comment (bparrishMines) [inline packages/firebase_ml_vision/example/android/app/build.gradle]: Does Auto ML Vision Edge require 21+ or is this for adding the `camera` plugin?

File: packages/firebase_ml_vision/example/ios/Runner.xcodeproj/project.pbxproj [hunk 3]
```diff
@@ -38,6 +38,7 @@
 /* End PBXCopyFilesBuildPhase section */
```
Comment (bparrishMines) [inline packages/firebase_ml_vision/example/ios/Runner.xcodeproj/project.pbxproj]: You can remove this file from being checked in.

File: packages/firebase_ml_vision/example/lib/detector_painters.dart [hunk 4]
```diff
@@ -115,55 +106,103 @@ class LabelDetectorPainter extends CustomPainter {
...
 
   @override
   bool shouldRepaint(TextDetectorPainter oldDelegate) {
-    return oldDelegate.absoluteImageSize != absoluteImageSize ||
+    return oldDelegate.imageSize != imageSize ||
         oldDelegate.visionText != visionText;
   }
 }
+
+Rect _scaleRect({
```
Comment (bparrishMines) [inline packages/firebase_ml_vision/example/lib/detector_painters.dart]: Was this copied from the [ML Kit demo](https://github.com/bparrishMines/mlkit_demo). I recently archived that and the most updated code is inside this plugin. You should be able to uncheck this.

File: packages/firebase_ml_vision/example/lib/main.dart [hunk 5]
```diff
@@ -2,14 +2,13 @@
 // Use of this source code is governed by a BSD-style license that can be
```
Comment (bparrishMines) [inline packages/firebase_ml_vision/example/lib/main.dart]: I updated this to include three demos. You will most likely have to pull in the latest changes.

File: packages/firebase_ml_vision/example/lib/utils.dart [hunk 6]
```diff
@@ -0,0 +1,70 @@
+import 'dart:async';
```
Comment (bparrishMines) [inline packages/firebase_ml_vision/example/lib/utils.dart]: This file has now been added in a recent commit.

File: packages/firebase_ml_vision/example/test_driver/firebase_ml_vision.dart [hunk 7]
```diff
@@ -62,6 +62,21 @@ void main() {
       });
     });
 
+    group('$VisionEdgeImageLabeler', () {
```
Comment (bparrishMines) [inline packages/firebase_ml_vision/example/test_driver/firebase_ml_vision.dart]: Could you separate this to another file when you pull the latest changes in?

File: packages/firebase_ml_vision/ios/Classes/LocalVisionEdgeDetector.m [hunk 8]
```diff
@@ -0,0 +1,82 @@
+#import "FirebaseMlVisionPlugin.h"
```
Comment (bparrishMines) [inline packages/firebase_ml_vision/ios/Classes/LocalVisionEdgeDetector.m]: Just like the Java versions, you can probably share alot of the logic with the `ImageLabel.m`.

File: packages/firebase_ml_vision/android/src/main/java/io/flutter/plugins/firebasemlvision/LocalVisionEdgeDetector.java [hunk 9]
```diff
@@ -0,0 +1,104 @@
+package io.flutter.plugins.firebasemlvision;
```
Comment (bparrishMines) [inline packages/firebase_ml_vision/android/src/main/java/io/flutter/plugins/firebasemlvision/LocalVisionEdgeDetector.java]: So we now have a different structure for the `Detector`s on the native side. We changed them so the user could deallocate resources and have multiple running `Detector`s. You could probably share alot of the logic with the `ImageLabeler.java`.

File: packages/firebase_ml_vision/android/src/main/java/io/flutter/plugins/firebasemlvision/SetupLocalModel.java [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+  private SetupLocalModel() {}
+
+  @Override
+  public void setup(String modelName, MethodChannel.Result result) {
+    String finalPath = "flutter_assets/assets/" + modelName + "/manifest.json";
+    FirebaseLocalModel localModel = FirebaseModelManager.getInstance().getLocalModel(modelName);
+    if (localModel == null) {
+      localModel = new FirebaseLocalModel.Builder(modelName).setAssetFilePath(finalPath).build();
+      FirebaseModelManager.getInstance().registerLocalModel(localModel);
+      result.success("Model Setup Complete");
```
Comment (bparrishMines) [inline packages/firebase_ml_vision/android/src/main/java/io/flutter/plugins/firebasemlvision/SetupLocalModel.java]: Why do we return a string for this? Do we use it on the dart side.

File: packages/firebase_ml_vision/README.md [hunk 11]
```diff
@@ -8,9 +8,27 @@ For Flutter plugins for other Firebase products, see [FlutterFire.md](https://gi
 
 *Note*: This plugin is still under development, and some APIs might not be available yet. [Feedback](https://github.com/flutter/flutter/issues) and [Pull Requests](https://github.com/flutter/plugins/pulls) are most welcome!
 
+Note: AutoML Vision Edge: iOS Only
```
[the remark this answers is not in the document]
Comment (rishab2113) [reply packages/firebase_ml_vision/README.md]: This was leftover from a previous commit, it works for both Android and iOS as of now

File: packages/firebase_ml_vision/example/android/app/build.gradle (hunk 2 above)
Comment (rishab2113) [reply packages/firebase_ml_vision/example/android/app/build.gradle]: that was for the camera plugin - will revert!

File: packages/firebase_ml_vision/example/lib/main.dart (hunk 5 above)
Comment (rishab2113) [reply packages/firebase_ml_vision/example/lib/main.dart]: pulled in!

File: packages/firebase_ml_vision/example/test_driver/firebase_ml_vision.dart (hunk 7 above)
Comment (rishab2113) [reply packages/firebase_ml_vision/example/test_driver/firebase_ml_vision.dart]: will do

File: packages/firebase_ml_vision/android/src/main/java/io/flutter/plugins/firebasemlvision/SetupLocalModel.java (hunk 10 above)
Comment (rishab2113) [reply packages/firebase_ml_vision/android/src/main/java/io/flutter/plugins/firebasemlvision/SetupLocalModel.java]: The return type of the dart function is Future<String>, I set this up to make the output similar to SetupRemoteModel, which can throw errors

File: packages/firebase_ml_vision/example/ios/Runner.xcodeproj/project.pbxproj (hunk 3 above)
Comment (rishab2113) [reply packages/firebase_ml_vision/example/ios/Runner.xcodeproj/project.pbxproj]: removing this file fails the CI, saying that the application isn't configured for iOS, I've added it back for now

**Comment by rishab2113** [comment]
Hey @bparrishMines, I’ve fixed all the issues with the build - it seems like you’ve to sign a CLA. The last CI looks like it’s failing due to timeouts

**Comment by securingsincity** [comment]
is there a sense of when this will be incorporated into flutterfire? Should an issue be created there?

**Comment by bparrishMines** [comment]
@securingsincity Yes, if you create an issue for this, you should make it in https://github.com/firebaseextended/flutterfire.

I thought that I had already moved this PR to the new repo, but it looks like I haven't yet. I started a new [PR](https://github.com/FirebaseExtended/flutterfire/pull/1547) handling the transition.

Closing

**Pull request closed without merging**